model-optimizer/extensions/front/tf/KerasRNNReplacer.py

```diff
@@ -0,0 +1,237 @@
+    tensor_list_set_item_node = int_match['concatenation']
+    tensor_list_set_item_node_name = tensor_list_set_item_node.soft_get('name', tensor_list_set_item_node.id)
+    list_result_node = int_match['concatenation_result']
+
+    # 1. process the body graph to avoid unsupported operations: TensorListGetItem and TensorListSetItem
+    # replace TensorListGetItem with Squeeze node and iterate through slices using axis for input port
+    squeeze_list_element = create_op_with_const_inputs(body_graph, Squeeze, {1: int64_array(0)},
+                                                       {'name': 'TensorListGetItemSqueeze'})
+    tensor_list_get_item_node.in_port(0).get_connection().set_destination(squeeze_list_element.in_port(0))
+    tensor_list_get_item_node.out_port(0).get_connection().set_source(squeeze_list_element.out_port(0))
+    rename_nodes([(tensor_list_get_item_node, tensor_list_get_item_node_name + '/AbandonedName'),
+                  (squeeze_list_element, tensor_list_get_item_node_name)])
+    unstack_placeholder_layer_id = unstack_placeholder.internal_layer_id
+    Loop.update_port_map_value_ext(loop_node.input_port_map, 'internal_layer_id', unstack_placeholder_layer_id,
+                                   'axis', 0)
+
+    # replace TensorListSetItem with Unsqueeze and use axis attribute for corresponding Result node
+    # to concatenate results from different iterations
+    unsqueeze_list_element = create_op_with_const_inputs(body_graph, Unsqueeze, {1: int64_array(0)},
+                                                         {'name': 'TensorListSetItemUnsqueeze'})
+    tensor_list_set_item_node.in_port(2).get_connection().set_destination(unsqueeze_list_element.in_port(0))
+    tensor_list_set_item_node.out_port(0).get_connection().set_source(unsqueeze_list_element.out_port(0))
+    rename_nodes([(tensor_list_set_item_node, tensor_list_set_item_node_name + '/AbandonedName'),
+                  (unsqueeze_list_element, tensor_list_set_item_node_name)])
+    list_result_node_layer_id = list_result_node.internal_layer_id
+    Loop.update_port_map_value_ext(loop_node.output_port_map, 'internal_layer_id', list_result_node_layer_id,
+                                   'axis', 0)
+
+    # 2. process locality of Loop node in the main graph to avoid unsupported operations:
+    # TensorListFromTensor, TensorListReserve, and TensorListStack
+    # remove TensorListFromTensor and pass a tensor to Loop as is
+    unstack_node.out_port(0).get_connection().set_source(unstack_node.in_port(0).get_connection().get_source())
+
+    # remove TensorListStack to by-pass the node since the result from the Loop node is already concatenated
+    stack_node.out_port(0).get_connection().set_source(stack_node.in_port(0).get_connection().get_source())
+
+    # remove TensorListReserve generating Tensor container that is no longer needed
+    graph.remove_node(list_reserve_node.id)
```

Comment:
Why do we have so specific action here? Is this operation is removed during the graph cleanup?

Reply:
Removed.